docs/msi_v2/msi_with_credential_design.md

```diff
@@ -0,0 +1,295 @@
+
+## Source Detection Logic
+
+MSAL follows a source detection process to determine how to interact with MSI endpoints and acquire tokens.
+
+### Environment Variable Check
+
+MSAL checks for Azure resource type based on specific environment variables to determine if the application is running on:
+
+- **Service Fabric**
+- **App Service**
+- **Azure Arc**
+- **Cloud Shell**
+- **Machine Learning**
+
+If identified, MSAL will use the appropriate legacy MSI endpoint for that resource.
+
+### Fallback to IMDS
+
+- If no specific Azure resource is identified from the environment variables, MSAL will fall back to IMDS (VMs and VMSS).
+- In this new design, Before fully falling back to IMDS, MSAL will now **probe the Credential Endpoint in IMDS**.
+- MSAL probes to see if the `/credential` endpoint exists in IMDS.
+- If the `/credential` endpoint is unavailable, it falls back to the legacy `/token` endpoint.
+
+## MSI V2 /credential Endpoint Details
+
+### Short-Lived Credential Retrieval
+
+- The `/credential` endpoint provides a **temporary credential** instead of a direct access token.
+- This credential is only valid for a short duration (1 hour) and must be used **immediately** to acquire an access token from ESTS.
+- This mechanism improves security by reducing the lifetime of sensitive authentication materials.
+
+### Retry Logic
+
+MSAL uses the **default Managed Identity retry policy** for MSI V2 credntial/token requests, whether calling the ESTS endpoint or the new `/credential` endpoint.
+
+## Steps for MSI V2 Authentication
+
+This section outlines the necessary steps to acquire an access token using the MSI V2 `/credential` endpoint. 
+
+### 1. Check for an Existing (Platform) Certificate (Windows only)
+- Search for a specific certificate (`devicecert.mtlsauth.local`) in `Cert:\LocalMachine\My`.
+- If the certificate is not found in Local Machine, check Current User's certificate store (Cert:\CurrentUser\My).
+- If any certificate is found, extract its thumbprint and use it for authentication.
+
+### 2. Generate a New Certificate (if platform certificate is not found)
+- If no valid platform certificate is found in Cert:\LocalMachine\My or Cert:\CurrentUser\My, create a new in-memory self-signed certificate.
+- This applies especially to Linux VMs, where platform certificates are not pre-configured, and MSAL must always generate an in-memory certificate for MTLS authentication.
+
+#### Certificate Creation Requirements
+- **Subject Name:** CN=mtls-auth (subject name can be adjusted as needed).
+- **Validity Period:** 90 days.
+- **Key Export Policy:** Private key must be exportable to allow use for MTLS authentication.
+- **Key Usage must include:** Digital Signature, Key Encipherment and TLS Client Authentication.
+- **Storage:** The certificate should exist only in memory. It is not stored in the certificate store. It is discarded when the process exits.
```

Comment:
You mentioned to me we'll be caching it?

Reply:
yes

Reply:
You didn't update this. It's not "discarded", it's cached.